2023-11-29-diving-into-jtag-part-2.md

```diff
@@ -0,0 +1,218 @@
+* DATA[31:0] = `0x20000000`
+* A[3:2] = `0b01` (address of `TAR` register)
+* RnW = `0b0`
+
+![image](https://github.com/memfault/interrupt/assets/33341738/b5b2327b-9504-46f8-a69e-99ec34ce529b)
+
+6. We form a read request. To do this, we write to the `DRW` register:
+* DATA[31:0] = `0x00`
+* A[3:2] = `0b11` (address of `DRW` register)
+* RnW = `0b1`
+
+![image](https://github.com/memfault/interrupt/assets/33341738/1cdc3e8f-9334-48cb-838d-365104ad4352)
+
+7. Read the `DRW` register by reading 35 bits from the `DR` register.
+
+![image](https://github.com/memfault/interrupt/assets/33341738/0f26aae0-ffc5-42b5-b2fa-9a0b99f977e6)
+
+## Interacting with the Processor Core
+Well, we’ve explored how the debugger interacts with the microcontroller’s memory. Now, let’s delve into how the debugger can halt program execution, step through code, read the processor’s registers, and so on. In the case of the STM32F407VG microcontroller, this is achieved through special registers:
+
+![image](https://github.com/memfault/interrupt/assets/33341738/3b57fc44-aa8f-42d1-8efd-11547ad77dfb)
+
+As detailed in section *C1.6 Debug system registers* of the [ARMv7-M Architecture Reference Manual](https://developer.arm.com/documentation/ddi0403/latest/), these registers play a pivotal role in debugging.
+
+What’s crucial for us now is understanding that by setting the `C_DEBUGEN` and `C_HALT` bits of the `DHCSR` register, we can halt the execution of the program. Similarly, by setting the `C_DEBUGEN` and `C_STEP` bits, we can execute the program step by step.
+
+Access to these registers via the DAP module is essentially the same as accessing memory. Instead of providing a memory address, you provide the address of the desired register. For instance, in the case of the `DHCSR` register, the address is `0xE000EDF0`. If you let me, I won't delve into an example with the bit sequence since it would be identical to the example from the chapter on memory interaction.
+
+## Closing
+Well, we’ve briefly explored how JTAG is used in microcontroller debugging. This topic is big, and it’s challenging to cover it fully within a single article. In this piece, we focused on invasive debugging, while non-invasive debugging wasn’t even mentioned. However, that wasn’t the goal when writing this article. For those who want to delve deeper into this topic, I recommend studying the links list listed below. As for us, we’ll move on, and in the next article, we’ll tackle the subject of testing printed circuit boards using the JTAG protocol.
+
+Thank you for your time! :)
+
+Next part: [Diving into JTAG. Part 3 — Boundary Scan](https://medium.com/@aliaksandr.kavalchuk/diving-into-jtag-part-3-boundary-scan-17f9975ecc59)
```

Comment:
Let's remove the link to part 3. We can add a link to our version once it is published though! 

Also, let's backlink to the original article here like Blake did in the previous one

```suggestion
>**Note**: This article was originally published by Aliaksandr on his blog. You can find the original article [here](https://medium.com/@aliaksandr.kavalchuk/diving-into-jtag-protocol-part-2-debugging-56a566db3cf8)
```

Reply:
Removed the "Next Part" section from the file (edited directly) and then committed your other suggestion - hope I'm doing this right... taking screenshots of your original comments just in case! Thanks so much Gilly